### PR TITLE
Consider overridden/custom types for doc blocks to achieve correct code completion

### PR DESCRIPTION
**Pull request by jdreesen**
This is a continuation of #14233 from @ctippler, adding support for custom document types, as requested by @mattamon.

It also includes a partly backport of #13934 to minimize conflicts when `10.6` gets merged into `v11.x`.

### CODE REVIEW

File: models/DataObject/ClassDefinition/Data/Extension/Relation.php [hunk 1]
```diff
@@ -26,49 +35,88 @@ trait Relation
...
+                return $typeLoader->getClassNameFor($shortName);
+            } catch (UnsupportedException) {
+                // try next
+            }
+
+            try {
+                return $typeLoader->build($shortName)::class;
+            } catch (UnsupportedException) {
+                // try next
+            }
```
Comment (jdreesen) [inline models/DataObject/ClassDefinition/Data/Extension/Relation.php]: I'm not entirely sure if we should do this. What happens if the constructor has mandatory arguments? Could that happen?

**Comment by brusch** [comment]
@jdreesen Thanks a lot for picking up this topic! 
I'd like to put this on `11.x` milestone, since `10.6` is intended to be used only for v11 preparations 😊 Can you rebase your PR, please? Thanks!

**Comment by mattamon** [comment]
https://github.com/pimcore/pimcore/issues/14326 Please be aware that we already have an issue for this in 11 and a PR
https://github.com/pimcore/pimcore/pull/14407

**Comment by jdreesen** [comment]
Should I close this then?

**Comment by mattamon** [comment]
Maybe you could provide some suggestions and inputs to the https://github.com/pimcore/pimcore/pull/14407. That would be very much appreciated and then yes we could close this one.

**Comment by jdreesen** [comment]
I think #14407 is fine, if we don't have to consider `pimcore.model.factory` anymore.

**Comment by mattamon** [comment]
> I think #14407 is fine, if we don't have to consider `pimcore.model.factory` anymore.

We well add the `pimcore.model.factory` to be safe.

**Comment by jdreesen** [comment]
I reverted the last two commits. Now it's not as forward compatible (it probably won't be merged anyway, so it doesn't matter), but it's easier to see the actual implementation.

**Comment by mattamon** [comment]
@jdreesen thanks a lot :) you are right, but I will close it now. Thanks again!

**Comment by jdreesen** [comment]
That's okay. I just wanted to make it easier if #14407 wants to compare with what I did here.

**Pull request closed without merging**